vwifi.c

```diff
@@ -680,6 +681,181 @@ static struct net_device_ops owl_ndev_ops = {
+
+    /* Add vif into global vif_list */
+    if (mutex_lock_interruptible(&owl->lock))
+        goto l_error_add_list;
+    list_add_tail(&vif->list, &owl->vif_list);
+    mutex_unlock(&owl->lock);
+
+    return &vif->wdev;
+
+l_error_add_list:
+    unregister_netdev(vif->ndev);
+l_error_ndev_register:
+    free_netdev(vif->ndev);
+l_error_alloc_ndev:
+    return NULL;
+}
+
+/* Unregister and free a virtual interface identified by @vif->ndev.
+ * This function could be called in cfg80211 context or
+ * by owl_exit() which is out of cfg80211 context.
+ */
+static int owl_delete_interface(struct owl_context *owl,
+                                struct owl_vif *vif,
+                                bool cfg80211_ctx)
+{
+    struct owl_packet *pkt = NULL, *safe = NULL;
+
+    if (mutex_lock_interruptible(&owl->lock))
+        return -ERESTARTSYS;
+
+    list_del(&vif->list);
+    mutex_unlock(&owl->lock);
+
+    /* Stop TX queue, and delete the pending packets */
+    netif_stop_queue(vif->ndev);
+    list_for_each_entry_safe (pkt, safe, &vif->rx_queue, list) {
+        list_del(&pkt->list);
+        kfree(pkt);
+    }
+    /* If there's is a pending scan, call cfg80211_scan_done to finish it. */
+    if (vif->scan_request) {
+        struct cfg80211_scan_info info = {
+            .aborted = true,
+        };
+
+        cfg80211_scan_done(vif->scan_request, &info);
+        vif->scan_request = NULL;
+    }
+
+    /* Make sure that no work is queued */
+    del_timer_sync(&vif->scan_timeout);
+    cancel_work_sync(&vif->ws_connect);
+    cancel_work_sync(&vif->ws_disconnect);
+    cancel_work_sync(&vif->ws_scan);
+    cancel_work_sync(&vif->ws_scan_timeout);
+
+    /* In contexts coming from cfg80211 callbacks, you must call this rather
+     * than unregister_netdevice(), unregister_netdev() is impossible as the
+     * RTNL is held. Otherwise, both unregister_netdevice() and
+     * unregister_netdev() are usable instead as well.
+     */
+    if (cfg80211_ctx)
+        cfg80211_unregister_netdevice(vif->ndev);
+    else
+        unregister_netdev(vif->ndev);
+    free_netdev(vif->ndev);
+
+    return 0;
+}
+
+/* Callback for unregistering and freeing the virtual
+ * interface identified by @wdev.
+ * The real deleteing work is offloaded to owl_delete_interface().
+ */
+static int owl_del_iface(struct wiphy *wiphy, struct wireless_dev *wdev)
+{
+    struct owl_context *owl = wiphy_get_owl_context(wiphy);
+    struct owl_vif *vif = wdev_get_owl_vif(wdev);
+
+    return owl_delete_interface(owl, vif, true);
+}
+
+/* Structure of functions for FullMAC 80211 drivers.
+ * Functions implemented along with fields/flags in wiphy structure would
+ * represent drivers features. This module can only perform "scan" and
+ * "connect". Some functions cant be implemented alone, for example: with
+ * "connect" there is should be function "disconnect".
+ */
+static struct cfg80211_ops owl_cfg_ops = {
+    .add_virtual_intf = owl_add_iface,
```

Comment:
Good idea to use the virtual interface. Could we directly assign the `owl0` or `owl1` to other `netns` ?
If it could work, we could remove the `macvlan`.

Reply:
Ok. I'll try to put every new virtual interface to a standalone `netns` in `cfg80211_ops->add_iface`. For naming convention, I think we could name `owl0_ns` if the interface name is `owl`.

Reply:
I tried to put owl0 to a namespace, and I got a error message:
```
$ sudo ip netns add owl0_ns
RTNETLINK answers: Invalid argument
```

Reply:
Also, move PHY interface to netns.
https://superuser.com/questions/653996/how-to-move-wireless-connection-to-other-network-namespace

Reply:
Ok. Currently all the interfaces created by iface_add will share the same PHY, so I have to create a PHY for every interface rather than share the same PHY.